### PR TITLE
Allow x509.certificate_managed to use a CSR

### PR DESCRIPTION
**Pull request by alxwr**
This capability was present prior to e9e49d17d6f.

### What does this PR do?

Bring back signing of CSRs.

### What issues does this PR fix or reference?

I don't know of any reported issue.

### Previous Behavior

```
          ID: /path/to/my.crt                                                                                       
    Function: x509.certificate_managed                                                                                                                                          
      Result: False                                                                                                                                                             
     Comment: Attempt 1: Returned a result of "False", with the following comment: "An exception occurred in this state: Traceback (most recent call last):                     
                File "/usr/local/lib/python3.7/site-packages/salt/state.py", line 2154, in call                                                                                 
                  *cdata["args"], **cdata["kwargs"]                                                                                                                             
                File "/usr/local/lib/python3.7/site-packages/salt/loader.py", line 2085, in wrapper                                                                             
                  return f(*args, **kwargs)                                                                                                                                     
                File "/usr/local/lib/python3.7/site-packages/salt/states/x509.py", line 659, in certificate_managed                                                             
                  "public_key or signing_private_key must be specified."                                                                                                        
              salt.exceptions.SaltInvocationError: public_key or signing_private_key must be specified.
```

### New Behavior

CSR is signed. Certificate is created.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [X] Docs
  - The docs state/imply that signing a CSR is possible. So there was no change necessary.
- [X] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
No

----

@glynnforrest This may interest you.

### CODE REVIEW

File: salt/states/x509.py [hunk 1]
```diff
@@ -651,7 +649,11 @@ def certificate_managed(
...
         )
 
-    if "public_key" not in kwargs and "signing_private_key" not in kwargs:
+    if (
+        "public_key" not in kwargs
+        and "signing_private_key" not in kwargs
+        and "csr" not in kwargs
+    ):
         raise salt.exceptions.SaltInvocationError(
             "public_key or signing_private_key must be specified."
```
Comment (glynnforrest) [inline salt/states/x509.py:658]:
```suggestion
            "public_key, signing_private_key, or csr must be specified."
```

**Review by glynnforrest — CHANGES_REQUESTED** [review]
Thanks for the ping @alxwr, I'm sorry that https://github.com/saltstack/salt/commit/e9e49d17d6fbd986d23a71f1bb415dca30cb3011 broke things for you. The fix looks good, though I'd like to see the error message updated please - one of the goals of https://github.com/saltstack/salt/pull/52935 was to make the error messages clear and unambiguous. Cheers for adding a test.

I can't comment on the other formatting changes you've made, e.g. `{0}` -> `{}`.

**Comment by alxwr** [comment]
> Thanks for the ping @alxwr, I'm sorry that [e9e49d1](https://github.com/saltstack/salt/commit/e9e49d17d6fbd986d23a71f1bb415dca30cb3011) broke things for you.

np :-) Thanks for the refactoring and making things easier!

> The fix looks good, though I'd like to see the error message updated please - one of the goals of #52935 was to make the error messages clear and unambiguous.

I thought about that and then forgot it. :-) Thanks for the suggestion! I applied it right away.

> Cheers for adding a test.

Just recently learned about `nox`. Great tool!

> I can't comment on the other formatting changes you've made, e.g. `{0}` -> `{}`.

`pre-commit run` (as used in CI) made me do it. :-D
(see also https://github.com/saltstack/salt/pull/58272#issuecomment-678937015)

**Review by krionbsd — APPROVED** [review]

**Review by glynnforrest — APPROVED** [review]

**Comment by alxwr** [comment]
@krionbsd @sagetherage Is there a reason this is not merged? :-) How long do you expect me to keep this PR up-to-date with current developments in master?

**Comment by krionbsd** [comment]
> @krionbsd @sagetherage Is there a reason this is not merged? :-) How long do you expect me to keep this PR up-to-date with current developments in master?

sorry for delay, it's in the queue to be merged

**Comment by sagetherage** [comment]
thank you @krionbsd and yes, queued up for merging, we removed the label `Merge Ready` as it didn't mean exactly as it is worded so not to confuse, we are using more dynamic indicators for merge status and looking to be explicit and automatic instead. Apologies for the confusion and thank you for the mention! @alxwr

**Pull request merged**